File: app/src/main/java/com/sdv/lichnoti/ShiftCalculator.kt
package com.sdv.lichnoti

import java.time.LocalDate

object ShiftCalculator {

    enum class ShiftType(val code: String, val label: String, val emoji: String) {
        NGAY("NGAY", "Ngày", "☀️"),
        DEM("DEM", "Đêm", "🌙"),
        NGHI("NGHI", "Nghỉ", "😴")
    }

    data class CrewInfo(
        val id: String,
        val name: String,
        val offset: Int
    )

    // 12-day cycle: 4 Day, 2 Off, 4 Night, 2 Off
    private val CYCLE = arrayOf(
        ShiftType.NGAY, ShiftType.NGAY, ShiftType.NGAY, ShiftType.NGAY,
        ShiftType.NGHI, ShiftType.NGHI,
        ShiftType.DEM, ShiftType.DEM, ShiftType.DEM, ShiftType.DEM,
        ShiftType.NGHI, ShiftType.NGHI
    )

    // Anchor date: January 3, 2026
    private val ANCHOR_DATE = LocalDate.of(2026, 1, 3)
    private val ANCHOR_JDN = toJulianDayNumber(ANCHOR_DATE)

    // Crew definitions with offsets from lichkipsdv.com
    val CREWS = listOf(
        CrewInfo("A", "Kíp A", 0),
        CrewInfo("B", "Kíp B", 8),
        CrewInfo("C", "Kíp C", 4),
        CrewInfo("HC", "Hành Chính", -1)
    )

    data class ShiftInfo(
        val type: ShiftType,
        val isHoliday: Boolean,
        val holidayName: String?
    )

    data class HOStats(
        val total: Int,
        val remaining: Int
    )

    /**
     * Kiểm tra Thứ Bảy có phải ngày HO hay không theo 3 giai đoạn:
     * - Trước tháng 7/2026: MỌI Thứ Bảy đều là HO
     * - Tháng 7-12/2026: Thứ Bảy CÁCH TUẦN là HO
     * - Từ tháng 1/2027: Không còn Thứ Bảy HO
     */
    private fun isSaturdayHO(date: LocalDate): Boolean {
        if (date.dayOfWeek.value != 6) return false
        val y = date.year
        val m = date.monthValue
        return when {
            y > 2026 -> false                    // Từ 2027: không có T7 HO
            y == 2026 && m >= 7 -> {              // 7-12/2026: T7 cách tuần
                val diff = toJulianDayNumber(date) - ANCHOR_JDN
                mod(diff / 7, 2) == 0
            }
            else -> true                          // Trước 7/2026: mọi T7 đều HO
        }
    }

    /**
     * Tính toán thống kê số ngày HO của một kíp trong năm.
     * Khớp 100% logic trang gốc:
     * - Ngày lễ: chỉ tính HO cho kíp đi làm ca Ngày (ca gốc Ngày).
     * - Ngày thường: tính HO cho kíp đi làm (Ngày/Đêm) trùng Chủ Nhật hoặc Thứ Bảy HO.
     */
    fun getHOStatsForYear(crewId: String, year: Int, today: LocalDate): HOStats {
        if (crewId == "HC") return HOStats(0, 0)
        
        var total = 0
        var remaining = 0
        val startDate = LocalDate.of(year, 1, 1)
        val endDate = LocalDate.of(year, 12, 31)
        
        var curr = startDate
        while (!curr.isAfter(endDate)) {
            val isOfficialHol = isHoliday(curr)
            val actualShift = getActualShift(crewId, curr)
            
            val isSunday = curr.dayOfWeek.value == 7
            val isSatWorkHO = isSaturdayHO(curr)

            val isHOReal = if (isSunday || isSatWorkHO) {
                if (isOfficialHol) {
                    getShift(crewId, curr) == ShiftType.NGAY
                } else {
                    actualShift != ShiftType.NGHI
                }
            } else {
                false
            }

            if (isHOReal) {
                total++
                if (!curr.isBefore(today)) {
                    remaining++
                }
            }
            curr = curr.plusDays(1)
        }
        return HOStats(total, remaining)
    }

    /**
     * Lấy tổng số ngày lễ chính thức trong năm
     */
    fun getHolidayCountForYear(year: Int): Int {
        var count = 0
        val startDate = LocalDate.of(year, 1, 1)
        val endDate = LocalDate.of(year, 12, 31)
        var curr = startDate
        while (!curr.isAfter(endDate)) {
            if (isHoliday(curr)) {
                count++
            }
            curr = curr.plusDays(1)
        }
        return count
    }

    /**
     * Get the shift type for a specific crew on a specific date.
     */
    fun getShift(crewId: String, date: LocalDate): ShiftType {
        // HC = Hành Chính: Mon-Fri = Day, Sat-Sun = Off
        if (crewId == "HC") {
            return if (date.dayOfWeek.value <= 5) ShiftType.NGAY else ShiftType.NGHI
        }

        val crew = CREWS.find { it.id == crewId } ?: return ShiftType.NGHI
        val targetJDN = toJulianDayNumber(date)
        val diff = targetJDN - ANCHOR_JDN
        val holidayAdj = countHolidaysBetween(ANCHOR_DATE, date)
        val idx = mod(crew.offset + diff - holidayAdj, CYCLE.size)
        return CYCLE[idx]
    }

    /**
     * Get the actual shift type for a specific crew on a specific date,
     * taking official holidays into account.
     * On official holidays:
     * - HC is off.
     * - Only crews with base shift NGAY work Day shift (HO).
     * - Crews with base shift DEM or NGHI are off.
     */
    fun getActualShift(crewId: String, date: LocalDate): ShiftType {
        val baseShift = getShift(crewId, date)
        if (isHoliday(date)) {
            if (crewId == "HC") return ShiftType.NGHI
            if (isLunarNewYear(date)) return ShiftType.NGHI
            return baseShift
        }
        return baseShift
    }

    /**
     * Get detailed shift info including holiday status and ca lam.
     * Cứ đi làm vào Chủ Nhật thì là ngày HO.
     * Đi làm vào Thứ Bảy cách tuần (bắt đầu từ thứ Bảy 3/1/2026) cũng là ngày HO.
     * Các ngày khác đi làm chỉ là HO nếu trùng vào ngày lễ chính thức.
     */
    fun getShiftInfo(crewId: String, date: LocalDate): ShiftInfo {
        val actualShift = getActualShift(crewId, date)
        val isOfficialHol = isHoliday(date)
        val holName = getHolidayName(date)

        // Đi làm vào Chủ Nhật (dayOfWeek.value == 7)
        val isSundayWork = (date.dayOfWeek.value == 7) && (actualShift != ShiftType.NGHI)

        // Đi làm vào Thứ Bảy HO (theo 3 giai đoạn: trước 7/2026 mọi T7, 7-12/2026 T7 cách tuần, từ 2027 không)
        val isSaturdayWork = isSaturdayHO(date) && (actualShift != ShiftType.NGHI)

        // Ngày được coi là ngày HO hoặc lễ nếu: trùng lễ chính thức, hoặc đi làm Chủ Nhật, hoặc đi làm Thứ Bảy cách tuần
        val isHOorHoliday = isOfficialHol || isSundayWork || isSaturdayWork

        val finalHolName = when {
            isOfficialHol -> holName
            isSundayWork -> "Chủ Nhật đi làm (HO)"
            isSaturdayWork -> "Thứ Bảy đi làm (HO)"
            else -> null
        }

        return ShiftInfo(actualShift, isHOorHoliday, finalHolName)
    }

    /**
     * Check if a date is a public holiday.
     * Hỗ trợ động mọi năm cho Tết Dương Lịch, Tết Nguyên Đán, Giỗ Tổ Hùng Vương,
     * 30/4, 1/5, Quốc Khánh (1/9, 2/9), Ngày thành lập (24/11), và ngày cuối năm SDV Day (31/12).
     */
    fun isHoliday(date: LocalDate): Boolean {
        val y = date.year
        val m = date.monthValue
        val d = date.dayOfMonth

        // 1. Các ngày lễ dương lịch cố định hàng năm
        if (m == 1 && d == 1) return true   // Tết Dương lịch
        if (m == 4 && d == 30) return true  // Giải phóng miền Nam
        if (m == 5 && d == 1) return true   // Quốc tế Lao động
        if (m == 9 && (d == 1 || d == 2)) return true // Quốc khánh
        if (m == 11 && d == 24) return true // Kỷ niệm ngày thành lập công ty (hiển thị nhãn Lễ)
        if (m == 12 && d == 31) return true // SDV Day (ngày cuối của mọi năm, hiển thị nhãn SDV)

        // 2. Các ngày lễ âm lịch theo năm (Tết Nguyên Đán và Giỗ Tổ Hùng Vương)
        return when (y) {
            2025 -> {
                (m == 1 && d in 28..31) || (m == 2 && d == 1) || (m == 4 && d == 7)
            }
            2026 -> {
                (m == 2 && d in 16..20) || (m == 4 && d == 26)
            }
            2027 -> {
                (m == 2 && d in 5..9) || (m == 4 && d == 15)
            }
            2028 -> {
                (m == 1 && d in 25..29) || (m == 4 && d == 4)
            }
            2029 -> {
                (m == 2 && d in 12..16) || (m == 4 && d == 23)
            }
            else -> false
        }
    }

    /**
     * Kiểm tra xem ngày có thuộc 5 ngày nghỉ Tết Âm lịch hàng năm hay không.
     */
    fun isLunarNewYear(date: LocalDate): Boolean {
        val y = date.year
        val m = date.monthValue
        val d = date.dayOfMonth
        return when (y) {
            2025 -> (m == 1 && d in 28..31) || (m == 2 && d == 1)
            2026 -> m == 2 && d in 16..20
            2027 -> m == 2 && d in 5..9
            2028 -> m == 1 && d in 25..29
            2029 -> m == 2 && d in 12..16
            else -> false
        }
    }

    /**
     * Get holiday name if the date is a holiday.
     */
    fun getHolidayName(date: LocalDate): String? {
        val y = date.year
        val m = date.monthValue
        val d = date.dayOfMonth

        return when {
            m == 1 && d == 1 -> "Tết Dương lịch"
            m == 4 && d == 30 -> "Giải phóng miền Nam"
            m == 5 && d == 1 -> "Quốc tế Lao động"
            m == 9 && (d == 1 || d == 2) -> "Quốc khánh"
            m == 11 && d == 24 -> "Kỷ niệm ngày thành lập" // Bỏ chữ SDV để hiển thị nhãn "Lễ"
            m == 12 && d == 31 -> "SDV Day"
            // Âm lịch
            y == 2025 && ((m == 1 && d in 28..31) || (m == 2 && d == 1)) -> "Tết Nguyên Đán"
            y == 2025 && m == 4 && d == 7 -> "Giỗ Tổ Hùng Vương"
            
            y == 2026 && (m == 2 && d in 16..20) -> "Tết Nguyên Đán"
            y == 2026 && m == 4 && d == 26 -> "Giỗ Tổ Hùng Vương"
            
            y == 2027 && (m == 2 && d in 5..9) -> "Tết Nguyên Đán"
            y == 2027 && m == 4 && d == 15 -> "Giỗ Tổ Hùng Vương"
            
            y == 2028 && (m == 1 && d in 25..29) -> "Tết Nguyên Đán"
            y == 2028 && m == 4 && d == 4 -> "Giỗ Tổ Hùng Vương"
            
            y == 2029 && (m == 2 && d in 12..16) -> "Tết Nguyên Đán"
            y == 2029 && m == 4 && d == 23 -> "Giỗ Tổ Hùng Vương"
            
            isHoliday(date) -> "Ngày lễ"
            else -> null
        }
    }

    /**
     * Count holidays strictly between two LocalDates using simple loops (extremely safe).
     */
    private fun countHolidaysBetween(anchorDate: LocalDate, targetDate: LocalDate): Int {
        var count = 0
        if (targetDate.isAfter(anchorDate)) {
            var curr = anchorDate.plusDays(1)
            while (curr.isBefore(targetDate)) {
                if (isHoliday(curr)) {
                    count++
                }
                curr = curr.plusDays(1)
            }
        } else if (targetDate.isBefore(anchorDate)) {
            var curr = targetDate.plusDays(1)
            while (curr.isBefore(anchorDate) || curr == anchorDate) {
                if (isHoliday(curr)) {
                    count--
                }
                curr = curr.plusDays(1)
            }
        }
        return count
    }

    /**
     * Convert a LocalDate to Julian Day Number.
     */
    private fun toJulianDayNumber(date: LocalDate): Int {
        val y = date.year
        val m = date.monthValue
        val d = date.dayOfMonth
        val a = (14 - m) / 12
        val yy = y + 4800 - a
        val mm = m + 12 * a - 3
        return d + (153 * mm + 2) / 5 + 365 * yy + yy / 4 - yy / 100 + yy / 400 - 32045
    }

    private fun mod(a: Int, b: Int): Int {
        return ((a % b) + b) % b
    }

    /**
     * Find the next working day (NGAY or DEM) for a crew starting from a given date.
     */
    fun findNextWorkDay(crewId: String, fromDate: LocalDate): Pair<LocalDate, ShiftType> {
        var date = fromDate
        for (i in 0..30) { // Look ahead up to 30 days
            val shift = getShift(crewId, date)
            if (shift != ShiftType.NGHI) {
                return Pair(date, shift)
            }
            date = date.plusDays(1)
        }
        return Pair(fromDate, ShiftType.NGHI)
    }
}

File: docs/shift-calculator.js
/**
 * ShiftCalculator - Port trực tiếp từ ShiftCalculator.kt
 * Logic tính ca kíp SDV cho web
 */
const ShiftCalculator = (() => {
    const ShiftType = { NGAY: 'NGAY', DEM: 'DEM', NGHI: 'NGHI' };
    const SHIFT_LABELS = { NGAY: 'Ngày', DEM: 'Đêm', NGHI: 'Nghỉ' };
    const CYCLE = [
        ShiftType.NGAY, ShiftType.NGAY, ShiftType.NGAY, ShiftType.NGAY,
        ShiftType.NGHI, ShiftType.NGHI,
        ShiftType.DEM, ShiftType.DEM, ShiftType.DEM, ShiftType.DEM,
        ShiftType.NGHI, ShiftType.NGHI
    ];
    const CREWS = [
        { id: 'A', name: 'Kíp A', offset: 0 },
        { id: 'B', name: 'Kíp B', offset: 8 },
        { id: 'C', name: 'Kíp C', offset: 4 },
        { id: 'HC', name: 'Hành Chính', offset: -1 }
    ];

    // Anchor: Jan 3, 2026
    const ANCHOR = new Date(2026, 0, 3);

    function toJDN(date) {
        const y = date.getFullYear(), m = date.getMonth() + 1, d = date.getDate();
        const a = Math.floor((14 - m) / 12);
        const yy = y + 4800 - a;
        const mm = m + 12 * a - 3;
        return d + Math.floor((153 * mm + 2) / 5) + 365 * yy + Math.floor(yy / 4) - Math.floor(yy / 100) + Math.floor(yy / 400) - 32045;
    }

    function mod(a, b) { return ((a % b) + b) % b; }

    function dateEqual(a, b) {
        return a.getFullYear() === b.getFullYear() && a.getMonth() === b.getMonth() && a.getDate() === b.getDate();
    }

    function addDays(date, n) {
        const d = new Date(date);
        d.setDate(d.getDate() + n);
        return d;
    }

    function dayOfWeek(date) {
        const d = date.getDay();
        return d === 0 ? 7 : d; // 1=Mon..7=Sun
    }

    function isSaturdayHO(date) {
        if (dayOfWeek(date) !== 6) return false;
        const y = date.getFullYear(), m = date.getMonth() + 1, d = date.getDate();
        if (y > 2026) return false;
        if (y === 2026) {
            return (m===1&&(d===3||d===17))||(m===2&&d===14)||(m===3&&(d===14||d===28))||
                   (m===4&&(d===11||d===25))||(m===5&&(d===9||d===23))||(m===6&&(d===6||d===20))||
                   (m===7&&d===4)||(m===8&&(d===15||d===29))||(m===9&&(d===12||d===26))||
                   (m===10&&d===24)||(m===11&&d===7)||(m===12&&d===5);
        }
        const nth = Math.floor((d - 1) / 7) + 1;
        return nth === 1 || nth === 3;
    }

    function isHoliday(date) {
        const y = date.getFullYear(), m = date.getMonth() + 1, d = date.getDate();
        if (m===1&&d===1) return true;
        if (m===4&&d===30) return true;
        if (m===5&&d===1) return true;
        if (m===9&&(d===1||d===2)) return true;
        if (m===11&&d===24) return true;
        if (m===12&&d===31) return true;
        switch (y) {
            case 2025: return (m===1&&d>=28&&d<=31)||(m===2&&d===1)||(m===4&&d===7);
            case 2026: return (m===2&&d>=16&&d<=20)||(m===4&&d===26);
            case 2027: return (m===2&&d>=5&&d<=9)||(m===4&&d===16);
            case 2028: return (m===1&&d>=25&&d<=29)||(m===4&&d===4);
            case 2029: return (m===2&&d>=12&&d<=16)||(m===4&&d===23);
            default: return false;
        }
    }

    function isLunarNewYear(date) {
        const y = date.getFullYear(), m = date.getMonth() + 1, d = date.getDate();
        switch (y) {
            case 2025: return (m===1&&d>=28&&d<=31)||(m===2&&d===1);
            case 2026: return m===2&&d>=16&&d<=20;
            case 2027: return m===2&&d>=5&&d<=9;
            case 2028: return m===1&&d>=25&&d<=29;
            case 2029: return m===2&&d>=12&&d<=16;
            default: return false;
        }
    }

    function getHolidayName(date) {
        const y = date.getFullYear(), m = date.getMonth() + 1, d = date.getDate();
        if (m===1&&d===1) return 'Tết Dương lịch';
        if (m===4&&d===30) return 'Giải phóng miền Nam';
        if (m===5&&d===1) return 'Quốc tế Lao động';
        if (m===9&&(d===1||d===2)) return 'Quốc khánh';
        if (m===11&&d===24) return 'Kỷ niệm ngày thành lập';
        if (m===12&&d===31) return 'SDV Day';
        if (y===2025&&((m===1&&d>=28)||(m===2&&d===1))) return 'Tết Nguyên Đán';
        if (y===2025&&m===4&&d===7) return 'Giỗ Tổ Hùng Vương';
        if (y===2026&&m===2&&d>=16&&d<=20) return 'Tết Nguyên Đán';
        if (y===2026&&m===4&&d===26) return 'Giỗ Tổ Hùng Vương';
        if (y===2027&&m===2&&d>=5&&d<=9) return 'Tết Nguyên Đán';
        if (y===2027&&m===4&&d===16) return 'Giỗ Tổ Hùng Vương';
        if (y===2028&&m===1&&d>=25&&d<=29) return 'Tết Nguyên Đán';
        if (y===2028&&m===4&&d===4) return 'Giỗ Tổ Hùng Vương';
        if (y===2029&&m===2&&d>=12&&d<=16) return 'Tết Nguyên Đán';
        if (y===2029&&m===4&&d===23) return 'Giỗ Tổ Hùng Vương';
        if (isHoliday(date)) return 'Ngày lễ';
        return null;
    }

    function countHolidaysBetween(anchor, target) {
        let count = 0;
        if (target > anchor) {
            let curr = addDays(anchor, 1);
            while (curr < target) {
                if (isHoliday(curr)) count++;
                curr = addDays(curr, 1);
            }
        } else if (target < anchor) {
            let curr = addDays(target, 1);
            while (curr <= anchor) {
                if (isHoliday(curr)) count--;
                curr = addDays(curr, 1);
            }
        }
        return count;
    }

    function getShift(crewId, date) {
        if (crewId === 'HC') {
            return dayOfWeek(date) <= 5 ? ShiftType.NGAY : ShiftType.NGHI;
        }
        const crew = CREWS.find(c => c.id === crewId);
        if (!crew) return ShiftType.NGHI;
        const diff = toJDN(date) - toJDN(ANCHOR);
        const holidayAdj = countHolidaysBetween(ANCHOR, date);
        const idx = mod(crew.offset + diff - holidayAdj, CYCLE.length);
        return CYCLE[idx];
    }

    // Bảng ca trực lễ năm 2026 thực tế
    const REAL_2026_HOLIDAYS = {
        '1-1': { A: ShiftType.DEM, B: ShiftType.NGHI, C: ShiftType.NGAY },
        '4-26': { A: ShiftType.NGAY, B: ShiftType.NGHI, C: ShiftType.DEM },
        '4-30': { A: ShiftType.DEM, B: ShiftType.NGAY, C: ShiftType.NGHI },
        '5-1': { A: ShiftType.NGHI, B: ShiftType.NGAY, C: ShiftType.DEM },
        '9-1': { A: ShiftType.NGAY, B: ShiftType.DEM, C: ShiftType.NGHI },
        '9-2': { A: ShiftType.NGAY, B: ShiftType.NGHI, C: ShiftType.DEM },
        '11-24': { A: ShiftType.NGHI, B: ShiftType.NGAY, C: ShiftType.DEM },
        '12-31': { A: ShiftType.NGAY, B: ShiftType.DEM, C: ShiftType.NGHI }
    };

    function getActualShift(crewId, date) {
        if (crewId === 'HC') {
            return (isHoliday(date) || dayOfWeek(date) > 5) ? ShiftType.NGHI : ShiftType.NGAY;
        }
        if (isLunarNewYear(date)) return ShiftType.NGHI;
        if (isHoliday(date)) {
            const y = date.getFullYear(), m = date.getMonth() + 1, d = date.getDate();
            if (y === 2026) {
                const key = `${m}-${d}`;
                const map = REAL_2026_HOLIDAYS[key];
                if (map) return map[crewId] || ShiftType.NGHI;
            }
            const prevDay = addDays(date, -1);
            const nextDay = addDays(date, 1);
            const isPrevHol = isHoliday(prevDay) && !isLunarNewYear(prevDay);
            const isNextHol = isHoliday(nextDay) && !isLunarNewYear(nextDay);
            const baseShifts = { A: getShift('A', date), B: getShift('B', date), C: getShift('C', date) };
            const entries = Object.entries(baseShifts);
            const dayCrew = entries.find(e => e[1] === ShiftType.NGAY)?.[0];
            const nightCrew = entries.find(e => e[1] === ShiftType.DEM)?.[0];
            const offCrew = entries.find(e => e[1] === ShiftType.NGHI)?.[0];

            if (isNextHol) {
                if (crewId === offCrew) return ShiftType.NGAY;
                if (crewId === dayCrew) return ShiftType.DEM;
                if (crewId === nightCrew) return ShiftType.NGHI;
            } else if (isPrevHol) {
                const bsPrev = { A: getShift('A', prevDay), B: getShift('B', prevDay), C: getShift('C', prevDay) };
                const eP = Object.entries(bsPrev);
                const dc = eP.find(e => e[1] === ShiftType.NGAY)?.[0];
                const nc = eP.find(e => e[1] === ShiftType.DEM)?.[0];
                const oc = eP.find(e => e[1] === ShiftType.NGHI)?.[0];
                if (crewId === oc) return ShiftType.NGAY;
                if (crewId === dc) return ShiftType.NGHI;
                if (crewId === nc) return ShiftType.DEM;
            } else {
                if (crewId === dayCrew) return ShiftType.NGAY;
                if (crewId === nightCrew) return ShiftType.NGHI;
                if (crewId === offCrew) return ShiftType.DEM;
            }
            return ShiftType.NGHI;
        }
        return getShift(crewId, date);
    }

    function getShiftInfo(crewId, date) {
        const actualShift = getActualShift(crewId, date);
        const isOfficialHol = isHoliday(date);
        const holName = getHolidayName(date);
        const isSundayWork = dayOfWeek(date) === 7 && actualShift !== ShiftType.NGHI;
        const isSaturdayWork = isSaturdayHO(date) && actualShift !== ShiftType.NGHI;
        const isHO = isOfficialHol || isSundayWork || isSaturdayWork;
        let finalHolName = null;
        if (isOfficialHol) finalHolName = holName;
        else if (isSundayWork) finalHolName = 'Chủ Nhật đi làm (HO)';
        else if (isSaturdayWork) finalHolName = 'Thứ Bảy đi làm (HO)';
        return { type: actualShift, isHoliday: isHO, holidayName: finalHolName, isOfficialHoliday: isOfficialHol };
    }

    function getHOStats(crewId, year) {
        if (crewId === 'HC') return { total: 0, remaining: 0 };
        let total = 0, remaining = 0;
        const today = new Date(); today.setHours(0,0,0,0);
        let curr = new Date(year, 0, 1);
        const end = new Date(year, 11, 31);
        while (curr <= end) {
            const actualShift = getActualShift(crewId, curr);
            const isSunday = dayOfWeek(curr) === 7;
            const isSatWork = isSaturdayHO(curr);
            if ((isSunday || isSatWork) && actualShift !== ShiftType.NGHI) {
                total++;
                if (curr >= today) remaining++;
            }
            curr = addDays(curr, 1);
        }
        return { total, remaining };
    }

    return { ShiftType, SHIFT_LABELS, CREWS, getShiftInfo, getHOStats, isHoliday };
})();
